feat(suite-desktop): new way of launching processes

keraf opened the pull request:
- Fixes #2655.
- Addresses #2719.
- Required for #2445.

prusnak commented on packages/suite-desktop/src-electron/processes/BridgeProcess.ts:
```diff
@@ -0,0 +1,41 @@
+    constructor() {
+        super();
+
+        this.processName = `trezord-${getPlatformName()}-x64`;
```
```suggestion
        this.processName = `trezord`;
```

prusnak commented on packages/suite-desktop/src-electron/processes/BaseProcess.ts:
```diff
@@ -0,0 +1,86 @@
+        }
+
+        const extension = getPlatformExt();
+        const processPath = path.join(RESOURCES, this.resourceName, this.processName);
```
```suggestion
        const processPath = path.join(RESOURCES, 'bin', this.resourceName, this.processName);
```

szymonlesisz replied:
make sure that this `bin` is not platform specific (ex: windows)

prusnak replied:
> make sure that this `bin` is not platform specific (ex: windows)

No problem here, `bin` is our folder, not system on.

prusnak reviewed (changes requested):
Left two suggestions below (fallout from 35462c9d6e78052c6ff292c9c02d602d7a183c86).

I think we can simplify the code substantially if we:
* don't try to kill processes we have not started (this also means that we should drop the `force-bridge` option) - it can be argued whether this makes more harm than good, there are lots of cases when the Suite can't kill the bridge anyway, because it has been installed as a system service and it is thus being run under different user than Suite
* if we drop this option, we can also drop `getProcessId` + `psList` dependency (which sounds like a horrible hack) and just store the `processID` when we start the bridge (and kill the bridge by `processID`, not by `processName`, when we exit) - if we keep `processID` as null, we know that the bridge was not started by us and we should not kill it

I am changing this to draft to make sure no one merges this by accident before we address these (and maybe other) questions.

prusnak commented:
Also let's replace `getPlatformName` + `getPlatformExt` in `src-electron/utils/process.ts` with the `getPlatformArchExt` implementation I wrote in `src-electron/bridge.ts`. I like that it has the sanity check for `supportedSystems`.

szymonlesisz commented on packages/suite-desktop/src-electron/constants.ts:
```diff
@@ -0,0 +1,3 @@
+import isDev from 'electron-is-dev';
```
shouldnt this be a part of `BaseProcess` since it's used only there?

keraf replied:
It's used in `electron.ts` but forgot to reference it there.

szymonlesisz commented on packages/suite-desktop/src-electron/processes/BridgeProcess.ts:
```diff
@@ -0,0 +1,41 @@
+    async status(): Promise<Status> {
+        // service
+        try {
+            const resp = await fetch(`http://127.0.0.1:21325/status/`);
```
- fetch POST on "/" would be quicker (less data to receive)
- are you sure it's working? (is it sending valid origin to trezord and accepted by cors check),

keraf replied:
I pointed to /status/ in order to get a 200 from a known page of trezord. The root path redirects to this one. I could check for a redirect code or follow the redirection for the 200 code which would result to the same. Not sure if it's sending a valid origin but I get a 200 back :smile: I'll run more tests to make sure it isn't just luck.

szymonlesisz replied:
if you send POST request you will not be redirected and you should receive only trezord version back:
`{version: "2.0.29"}`
just for reference:
https://github.com/trezor/trezord-go#api-documentation

keraf replied:
Played around with POST on / and found out it requires a valid origin unlike /status. Updated in 49bc25c.

szymonlesisz commented on packages/suite-desktop/src-electron/electron.ts:
```diff
@@ -346,7 +349,11 @@ app.on('browser-window-focus', (_event, win) => {
+        if (devMode) {
+            await bridge.startDev(forceBridge);
+        } else {
+            await bridge.start(forceBridge);
```
starting bridge from a user request should be always forced regardless to flag, no?
because it means that something is not right/not working

keraf commented:
Thanks for the comments @szymonlesisz & @prusnak. I updated the PR with the suggested changes and some clean up. I removed the `os` and `ps-list` packages and it's now only running/killing it's child process. Removing the `detached` parameter from `spawn` also guarantees that the process will be killed when the application is closed.

prusnak commented on packages/suite-desktop/src-electron/processes/BaseProcess.ts:
```diff
@@ -0,0 +1,109 @@
+            RESOURCES,
+            'bin',
+            this.resourceName,
+            system,
```
the build described in `suite-desktop/package.json` takes care of copying the correct binary from `bin/${resourceName}/${system}/${processName}` to `resources/bin/${resourceName}/${processName}`

so the `system` is not part of the path in the final result image

```suggestion
```

keraf replied:
That seems to be the case for production builds only. I was looking at the output for dev builds :smile: I'll add a condition for dev/prod paths.

prusnak commented on packages/suite-desktop/src-electron/processes/BaseProcess.ts:
```diff
@@ -0,0 +1,109 @@
+     */
+    async stop() {
+        if (this.process) {
+            this.process.kill('SIGINT');
```
This is more standard way (on Unix, not sure about Windows :-/):

```suggestion
            if (!this.process.kill()) {
                // wait for 5 seconds
                this.process.kill('SIGKILL');
            }
```

prusnak replied:
Reopening - this has not been addressed yet.

keraf replied:
Electron runs on a version of node that unfortunately doesn't have kill return anything. I think it has been added on a later version of node 12.

prusnak reviewed (changes requested):
Much better! Left some smaller suggestions below.

keraf commented:
Blows up on prod builds for some reason. Switching back to draft and investigating. 
![image](https://user-images.githubusercontent.com/30026749/97573111-8edf3800-19e9-11eb-845b-66a487d9cd0b.png)

Edit: All good, this is now fixed! Subfolders weren't included in the build. Tweaking the `build.files` property in `package.json` did the trick.

prusnak reviewed (commented):
Good, almost there!

One lame question, though - isn't this better than the current way?

``` javascript
class BaseProcess {
    constructor(resourceName, processName) {
        this.process = null;
        this.resourceName = resourceName;
        this.processName = processName;
    }
}

class BridgeProcess extends BaseProcess {
    constructor() {
        super('bridge', 'trezord');
    }
}
```

or that's not how you do things in JS? (I agree the code above is influenced by how you write stuff in Python).

prusnak commented on packages/suite-desktop/src-electron/processes/BaseProcess.ts:
```diff
@@ -67,15 +67,21 @@ class BaseProcess {
+
+            setTimeout(() => {
+                if (!this.process.killed) {
+                    this.process.kill('SIGINT');
```
```suggestion
                    this.process.kill('SIGKILL');
```

prusnak commented on packages/suite-desktop/src-electron/processes/BaseProcess.ts:
```diff
@@ -67,15 +67,21 @@ class BaseProcess {
+                }
+                this.process = null;
+                resolve();
+            }, 3000);
```
```suggestion
            }, 10000);
```

keraf replied:
I'll rewrite this bit. If the process gets killed without forcing it, this will delay the resolution by X seconds for nothing.

prusnak commented on packages/suite-desktop/src-electron/electron.ts:
```diff
@@ -346,13 +346,23 @@ app.on('browser-window-focus', (_event, win) => {
             await bridge.start();
         }
+
+        // Throttle bridge start
```
Is throttling bridge-specific or should we abstract this into BaseProcess?

keraf replied:
The front-end is calling `bridge/start` multiple times until the bridge starts, so I guess we can say it's bridge specific for now. I'd like to fix that in the front-end (as well/instead) but as I was already working on the back-end part, I fixed it there :smile:

prusnak replied:
Ah, okay. In that case, let's leave it here :)

prusnak reviewed (approved):
Very nice! One question below.

prusnak reviewed (approved):
untested ACK!

szymonlesisz reviewed (approved):
looks great, didnt test it

prusnak commented:
@keraf Feel free to merge if the CI is green, thanks!

keraf commented:
[CI passed](https://gitlab.com/satoshilabs/trezor/trezor-suite/-/pipelines/209302484) before squashing the last fixup. It now failed because of some docker timeout.
